### PR TITLE
Refactor Media & Text to use functional components

### PR DESCRIPTION
**Pull request by oxyc**
## Description

Originally submitted together with light block DOM conversion but now split into it's own PR to only refactor the Media & Text block using functional components. Original PR is here https://github.com/WordPress/gutenberg/pull/23062

Note that there might be some prep here for rendering MediaContainer as a `<figure>` directly.

## How has this been tested?

Tested adding video as well as image and toggling all the attributes to see that everything is working.

## Types of changes

Refactor

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-library/src/media-text/edit.js [hunk 1]
```diff
@@ -122,228 +103,195 @@ class MediaTextEdit extends Component {
...
+		mediaWidth,
+		rel,
+		verticalAlignment,
+	} = attributes;
+
+	const image = useSelect(
+		( select ) => {
+			const { getMedia } = select( 'core' );
+			return {
+				image: mediaId && isSelected ? getMedia( mediaId ) : null,
```
Comment (ellatrix) [inline packages/block-library/src/media-text/edit.js:132]: If there's only one key, you can return the value itself.

File: packages/block-library/src/media-text/edit.js [hunk 2]
```diff
@@ -122,228 +103,195 @@ class MediaTextEdit extends Component {
...
+		mediaId,
+		mediaPosition,
+		mediaType,
+		mediaUrl,
+		mediaWidth,
+		rel,
+		verticalAlignment,
+	} = attributes;
+
+	const image = useSelect(
```
Comment (ellatrix) [inline packages/block-library/src/media-text/edit.js:128]: Shouldn't this be { image } if you return an object below?

File: packages/block-library/src/media-text/media-container.js [hunk 3]
```diff
@@ -43,141 +47,121 @@ function ResizableBoxContainer( { isSelected, isStackedOnMobile, ...props } ) {
...
+		};
 
-	renderToolbarEditButton() {
-		const { onSelectMedia, mediaUrl, mediaId } = this.props;
 		return (
-			<BlockControls>
-				<MediaReplaceFlow
+			<ResizableBoxContainer
+				className="editor-media-container__resizer"
+				style={ backgroundStyles }
```
Comment (oxyc) [inline packages/block-library/src/media-text/media-container.js:140]: Oops, stray from old PR

File: packages/block-library/src/media-text/media-container.js [hunk 4]
```diff
@@ -43,141 +47,121 @@ function ResizableBoxContainer( { isSelected, isStackedOnMobile, ...props } ) {
...
-					{ mediaElement }
-				</ResizableBoxContainer>
-			);
-		}
-		return this.renderPlaceholder();
-	}
+	return <PlaceholderContainer { ...props } />;
 }
 
 export default compose( [
```
Comment (ellatrix) [inline packages/block-library/src/media-text/media-container.js:167]: withDispatch can also be removed, right?

File: packages/block-library/src/media-text/edit.js (hunk 2 above)
Comment (oxyc) [reply packages/block-library/src/media-text/edit.js:128]: True, I must have forgotten to test it. I now confirmed that the `<ImageURLInputUI>` component which used the object works.

File: packages/block-library/src/media-text/edit.js [hunk 5]
```diff
@@ -56,22 +54,9 @@ const applyWidthConstraints = ( width ) =>
...
-		};
-		this.onSetHref = this.onSetHref.bind( this );
-	}
-
-	onSelectMedia( media ) {
-		const { setAttributes } = this.props;
-		const { linkDestination, href } = this.props.attributes;
+function attributesFromMedia( { attributes, setAttributes } ) {
+	return ( media ) => {
+		const { linkDestination, href } = attributes;
```
Comment (ZebulanStanphill) [inline packages/block-library/src/media-text/edit.js:59]: You could condense some of the destructuring here.
```suggestion
function attributesFromMedia( { attributes: { linkDestination, href }, setAttributes } ) {
	return ( media ) => {
```

**Review by ZebulanStanphill — APPROVED** [review]
Code looks good and testing the branch hasn't revealed any issues. Looks good to merge (LGTM)!

**Pull request merged**